pkg/component/etcd/leases.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"context"
+	"fmt"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/pkg/common"
+	"github.com/gardener/gardener/pkg/controllerutils"
+
+	"github.com/gardener/gardener/pkg/utils/flow"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (e *etcd) emptyLease(name string) *coordinationv1.Lease {
+	return &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: e.namespace,
+		},
+	}
+}
+
+func (e *etcd) reconcileLease(ctx context.Context, lease *coordinationv1.Lease, enabled bool) error {
+	if !enabled {
+		return client.IgnoreNotFound(e.client.Delete(ctx, lease))
+	}
+	_, err := controllerutils.MergePatchOrCreate(ctx, e.client, lease, func() error {
+		lease.OwnerReferences = getOwnerReferences(e.values)
+		return nil
+	})
+	return err
+}
+
+func (e *etcd) reconcileMemberLease(ctx context.Context) error {
```

Comment:
This method is not reconciling a Lease object, it's creating, updating, or deleting multiple Lease objects (as part of the reconcilation of an Etcd object). The actual reconciliation of the Leases will be performed by the respective controller. Let's use a name that matches more closely what this method is doing. I propose `deployOrDestroy` here though I guess `deploy`, `update` or `createOrUpdate` would also be ok - but not `reconcile` which has a very different meaning and is therefore confusing.

```suggestion
func (e *etcd) deployOrDestroyMemberLeases(ctx context.Context) error {
```